lumicks/pylake/scan.py

```diff
@@ -31,32 +32,134 @@ def __init__(self, name, file, start, stop, metadata):
+
+            # Early out if this color channel is missing (slicing will otherwise fail) and
+            # we need to handle this gracefully because of files with missing color channels
+            if not img.size:
+                return img
+
+            return img[tuple(slices)]
+
+        def timestamp_factory(_, reduce_timestamps):
+            ts = self._timestamps("timestamps", reduce_timestamps)
+            return ts[tuple(slices)]
+
+        result = copy(self)
+        result._image_factory = image_factory
+        result._timestamp_factory = timestamp_factory
+        result._pixelcount_factory = pixelcount_factory
+
+        # Force reconstruction number of frames now
+        result._metadata = self._metadata.with_num_frames(
+            len(range(*frame_axis.indices(self.num_frames))) if isinstance(frame_axis, slice) else 1
+        )
+
+        # The metadata has an ugly hack when the number of frames metadata is missing
+        # this means that zero cannot be expressed correctly at this time since it
+        # would trigger a reconstruction. For now, we raise on an empty scan.
+        if result._metadata.num_frames == 0:
+            raise NotImplementedError("Slice is empty.")
+
+        # Verify that none of the axis end up empty
+        resulting_shape = pixelcount_factory(None)
+        if np.any(np.asarray(resulting_shape) == 0):
+            raise NotImplementedError("Slice is empty.")
+
+        return result
+
     def __repr__(self):
         name = self.__class__.__name__
         return f"{name}(pixels=({self.pixels_per_line}, {self.lines_per_frame}))"
 
     def __getitem__(self, item):
         """All indexing is in frames"""
-        ts_ranges = self.frame_timestamp_ranges(include_dead_time=True)
 
-        if isinstance(item, slice):
-            if item.step is not None:
-                raise IndexError("Slice steps are not supported")
+        def check_item(item, allow_int):
+            """We don't allow slice steps, and we don't allow slicing with lists. Slicing with
+            single integer value is only allowed for frames."""
+            if isinstance(item, slice):
+                if item.step is not None:
+                    raise IndexError("Slice steps are not supported when indexing")
+                return item
+            else:
+                if not allow_int:
+                    raise IndexError("Scalar indexing is not supported for spatial coordinates")
 
-            sliced_ts_ranges = ts_ranges[item]
-            if not sliced_ts_ranges:
-                raise NotImplementedError("Slice is empty.")
+                if isinstance(item, int):
+                    return item
 
-            start, stop = sliced_ts_ranges[0][0], sliced_ts_ranges[-1][-1]
-            num_frames = len(sliced_ts_ranges)
-        else:
-            start, stop = ts_ranges[item]
-            num_frames = 1
+                raise IndexError("Only slicing by int or slice is allowed to select frames")
```

Comment:
Right, that's a good point. I restructured it a bit and produce the correct error message now.

```python
            if isinstance(item, slice):
                if item.step is not None:
                    raise IndexError("Slice steps are not supported when indexing")
                return item

            if isinstance(item, int):
                if not allow_int:
                    raise IndexError("Scalar indexing is not supported for spatial coordinates")
                else:
                    return item

            raise IndexError(f"Slicing by {type(item).__name__} is not allowed.")
```

I think code-wise it's also a bit less confusing, since now you immediately see which type is attempted at each block.